### ISSUE TITLE
Performance issues in huge files

### ISSUE THREAD

**Issue by HTD** [jacobslusser/ScintillaNET#76]
I need to process huge files with very custom syntax. So I made a fast parser and took 10MB file to test it. The parser worked as follows: the text is buffered (Buffer = Ctl.GetTextRange(..)), then it's parsed into virtual regions (2 x int struct), then styled - each distinct style is applied to editor control with Ctl.SetStyling. And the results: "Buffer loaded in: 0.1s, Parsed in: 7.1s, Syled in: 19.1s". Of course it's done only once, when you show the last page.

Why is applying styles so slow? I believe visual operations are not performed for invisible part, so most of the time is used for calculations. I'm afraid it's the evil UTF-8 internal buffer. Is there any hope it will be replaced with native UTF-16 support? Maybe a single offset map could be created on loading, and updated before huge reparse operations? Memory is cheap. BTW, what is 100MB in modern PC? Geez, even in ancient ones it's nothing. Basic byte shuffling operations are blazing fast in C#. Maybe not fast enough for real-time 3D graphics, but surely fast enough to handle Ctrl+End event on 10MB file.

First: is styling slow because UTF-16 / UTF-8 conversions? If so, I would try to address it by creating offset buffer for styling, but what's the best way to access native offsets and native methods? Or maybe ScintillaNET already does it?

There is another way to speed up displaying of huge files (I'm going for it right now). I'll create a new event "VisibleContentChanged" and connect my parser to it. This would unfortunately break automatic OnStyleNeeded() action. I tested it with quick and dirty code - it seems to work pretty fast (like real-time), however it will fail to correctly highlight regions larger than single screen content. It's a minor issue which can be solved by implementing new OnStyleNeeded() event based on virtual region maps.

Last but not least - for huge files I use BeginInvoke() and Application.DoEvents() triggered each 1KB - the performance impact is completely insignificant, but the application responsiveness is way better. You can view and edit the text with no problems while it's still finishing parsing and styling.

**Comment by jacobslusser** [comment]
> Why is applying styles so slow? I believe visual operations are not performed for invisible part, so most of the time is used for calculations.

I'm not entire sure that is a true statement. Depending on the style definitions, it's possible that applying a style--even to an unseen portion of the document--could require a redraw. This is because the text dimensions (height and width) could change from the styles used. Those measurements require a redraw to calculate the updated dimensions and scroll positions. This is just a hunch. To be sure you could attached Spy++ and watch the WM_PAINT messages.

> I'm afraid it's the evil UTF-8 internal buffer. Is there any hope it will be replaced with native UTF-16 support?

You can be sure that if I maintained Scintilla I would support UTF-16, but I only maintain the wrapper. It is what it is and internally Scintilla will continue to be UTF-8 for the foreseeable future.

> Maybe a single offset map could be created on loading, and updated before huge reparse operations? 

That's actually how the UTF-8 to UTF-16 conversion layer in ScintillaNET works. However, instead of maintaining a map of each byte/character offset, it keeps one for each line. This usually provides the best balance. A translation therefore usually first identifies the delta for the line, and then for the character within that line. I suppose it's possible to maintain a map for each character but that would likely cause a performance hit when the document changes. Sure the lookups would be fast, but a modification would require updating the offsets to each character.

> is styling slow because UTF-16 / UTF-8 conversions? If so, I would try to address it by creating offset buffer for styling, but what's the best way to access native offsets and native methods?

If for any reason you don't like the way ScintillaNET handles something you can use the `DirectMessage` API and native Scintilla [documentation](http://www.scintilla.org/ScintillaDoc.html) to sidestep it (i.e. `DirectMessage(SCI_STARTSTYLING, ...)`. It's not without risk, but you could directly interface with the underlying Scintilla control if you wanted to. Not knowing more about the issue I couldn't say whether you would get a performance increase or not by doing this. As I mentioned earlier, there is already a byte/character map in ScintillaNET to make lookups fast. If your performance problem is in styling, there is likely to be little gain because there is currently very little overhead in the ScintillaNET mapping lookup time; but I could be wrong....

> There is another way to speed up displaying of huge files (I'm going for it right now). I'll create a new event "VisibleContentChanged" and connect my parser to it.

That's a perfectly fine approach and definitely one I would recommend if you can. As you pointed out, it means that you might not know where in the document you are starting your styling and contextually sensitive styling might get messed up. The `StyleNeeded` event avoids this by playing catch-up, but if you don't need to then don't use that event.

If it is acceptable in your application to show moments of unstyled text, you could be very leisurely about how you approach styling. Once you're not using the `StyleNeeded` event, you could style the text at any pace/priority you want. Most editors take this approach and will show moments of unstyled text while the parser/styler catch up.

> for huge files I use BeginInvoke() and Application.DoEvents()

Without more context I couldn't say whether this is a good idea or not, but generally speaking using `Application.DoEvents()` is a bad idea. That might be a sign that something else is wrong in your application. You could inadvertently be preventing the Windows message loop from batching multiple paint events into one cycle.

Final thoughts... loading a 10MB file is not what most developers use Scintilla(NET) for and thus isn't something that is specifically optimized for. There could very well be a performance problem in ScintillaNET and if you want to help identify it I would be happy to work with you on a fix. But it might not be a "performance problem" at all. It might just be an unreal expectation that ScintillaNET is going to handle a 10MB file without using some tricks in your code to work around those design decisions/limitations.

**Comment by HTD** [comment]
Thank you for the great insight. I'm very glad the UTF-8 calculations are already made perfectly in ScintillaNET. You were right about Application.DoEvents() being no good in lexer code. I used UpdateUI event to apply lexer and now it works with 10MB files in real-time, no problems whatsoever.

I agree, it's not a performance issue, but I still can make some little enhancements to implement custom (Container) lexers easier, with various optimizations. I'll be glad to share my the code when it's ready.

(BTW, yesterday I took closer look at the source code, I'm impressed with the quality.)
